[RNMobile] Upgrade CI to use Xcode 12

ceyhun opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/master/docs/contributors/repository-management.md#pull-requests. -->



## Description
<!-- Please describe what you have changed or added -->

Fixes https://github.com/WordPress/gutenberg/issues/26188

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

- Check if all GitHub Actions workflows are green.
- Run e2e tests locally using Xcode 12:
    1. Make sure command line is using Xcode 12: `sudo xcode-select --switch /Applications/Xcode_12_path.app`
    1. `npm run native test:e2e:ios:local`

ceyhun committed:
Upgrade CI to use Xcode 12

ceyhun committed:
Invalidate iOS build cache

ceyhun committed:
Upgrade appium to 1.18.3

ceyhun committed:
Update appium config to use iOS 14.0 simulator

ceyhun committed:
Use iOS 13.4 simulator on Sauce Labs

ceyhun committed:
Fix lint error

guarani reviewed (commented):
When running the e2e tests locally I get:
```
error Failed to build iOS project. We ran "xcodebuild" command but it exited with error code 65. To debug build logs further, consider building your app with Xcode.app, by opening GutenbergDemo.xcworkspace. Run CLI with --verbose flag for more details.
```
Which could be issue https://github.com/WordPress/gutenberg/issues/25882, but which was fixed in PR https://github.com/WordPress/gutenberg/pull/26062 (and worked for me previously with this fix).

I tried running `npm run postinstall` as suggested in the above mentioned PR, but same result.
Then I switched back to the `master` branch and saw the same issue there as well when running `npm run native test:e2e:ios:local`. Notably, running `npm run native ios` on `master` succeeded, but didn't fire up the packager as it used to (running `npm run native start` in one window and `npm run native ios` in another window worked, and it's how I've been running Gutenberg for the last few months.

I tried:
1. Deleting `~/.cocoapods` and `cd packages/react-native-editor/ios && bundle exec pod install`
2. `cd packages/react-native-editor/ios && rm -rf Pods`, deleted Xcode derived data, and reran `npm run native test:e2e:ios:local` in the Gutenberg root folder

No luck. So I'll have to come back tomorrow for some more digging, but feel free to add another dev here to get a second look at this PR since this looks to be an issue with my machine, not this PR.

ceyhun committed:
Update ios e2e scripts to use xcodebuild directly, remove unused scripts

ceyhun committed:
Do not specify OS to use latest one

ceyhun committed:
Debug CI

ceyhun committed:
Invalidate iOS build cache

ceyhun committed:
Skip switching xcode

ceyhun committed:
Revert "Skip switching xcode"

This reverts commit a016323b34caa9c61d40f6907e89a1f0f9e48dcb.

ceyhun committed:
Install iOS dependencies before building iOS app

ceyhun committed:
Revert "Debug CI"

This reverts commit eac3886edee2db100dd89e0d5926c8414e2e3835.

ceyhun commented:
I think I reproduced the issue you were having. Was it maybe this one:
```
Can't find the 'node' binary to build the React Native bundle.  If you have a non-standard Node.js installation, select your project in Xcode, find  'Build Phases' - 'Bundle React Native code and images' and change NODE_BINARY to an  absolute path to your node executable. You can find it by invoking 'which node' in the terminal.
```

Strange that this happens only when using the `react-native run-ios` command to build and not when running from inside Xcode (even when using `FORCE_BUNDLING` flag) or when using `xcodebuild` directly. 

I updated the scripts to use `xcodebuild` directly and also removed some duplicate ones that we won't need.

For the metro server not launching on build: The `Build Phases` in the Xcode project doesn't include the phase to launch metro anymore, this I think was on purpose because the metro server needs to run from different roots depending on which gutenberg version is being executed (gutenberg or gutenberg-mobile one) and I'm not sure how to achieve that with the script living in the Xcode project.

ceyhun committed:
Merge branch 'master' into rnmobile/upgrade-ci-to-xcode-12

ceyhun committed:
Merge branch 'master' into rnmobile/upgrade-ci-to-xcode-12

guarani commented:
> I think I reproduced the issue you were having. Was it maybe this one:
> 
> ```
> Can't find the 'node' binary to build the React Native bundle.  If you have a non-standard Node.js installation, select your project in Xcode, find  'Build Phases' - 'Bundle React Native code and images' and change NODE_BINARY to an  absolute path to your node executable. You can find it by invoking 'which node' in the terminal.
> ```

I don't see that message in the output of `npm run native test:e2e:ios:local`, so it looks like it's a different issue.

> For the metro server not launching on build: The `Build Phases` in the Xcode project doesn't include the phase to launch metro anymore, this I think was on purpose because the metro server needs to run from different roots depending on which gutenberg version is being executed (gutenberg or gutenberg-mobile one) and I'm not sure how to achieve that with the script living in the Xcode project.

I see — I've been running the packager separately anyway, but good to know the reason why this changed.

---

I'm upgrading from Xcode 12.0 to Xcode 12.1 to see if that fixes the issue I'm having, since I'm still getting the same from https://github.com/WordPress/gutenberg/pull/26283#pullrequestreview-514303259 when using the latest commit from this branch.

ceyhun commented:
> I don't see that message in the output of npm run native test:e2e:ios:local, so it looks like it's a different issue.

@guarani Can you maybe share more details about the error, maybe the full output if that's possible?

@etoledom Can you please give this a try on your machine to see if you're having issues too?

etoledom commented on packages/react-native-editor/__device-tests__/helpers/caps.js:
```diff
@@ -1,12 +1,10 @@
 	deviceOrientation: 'portrait',
 	automationName: 'XCUITest',
-	appiumVersion: '1.16.0', // SauceLabs requires appiumVersion to be specified.
+	appiumVersion: '1.17.1', // Sauce Labs requires appiumVersion to be specified.
```
`appium` version on package.json is `1.18.3`. Is it expected to be different to the one declared here?

ceyhun replied:
`1.18.3` was not yet supported by Sauce Labs with iOS simulators: https://saucelabs.com/rest/v1/info/platforms/appium and older appium versions locally had issues with Xcode 12: https://github.com/appium/appium/issues/13532. As soon as Sauce Labs supports `1.18.3` with iOS we can bump this.

etoledom commented:
My first try with `npm run native test:e2e:ios:local` resulted on a error code 65 (even though the description was different)

Second try:
- `git clean -fdx`
- Xcode Clean Build Folder
- `npm i`
- `npm run native test:e2e:ios:local`

And Build Success! 🎉  
But all tests failed 😱 

Third try:
- I noticed that now the iOS version for local is `14.0`, I have Xcode 12.1 that comes with the `14.1` sims, so I changed the `platformVersion: '14.0'` def to `platformVersion: '14.1'` and now everything works! 🎉 🎉 

One thing, after doing the second try, I got a huge diff on `package-lock.json`. This could also be a difference on node version (or something like that).
Just to be sure:
```
$ node --version
v12.18.4
```

ceyhun commented:
Thanks for checking @etoledom, awesome that it worked 🎉

> One thing, after doing the second try, I got a huge diff on `package-lock.json`. This could also be a difference on node version (or something like that).

I couldn't reproduce this. I have node v12.19.0 and npm 6.14.8. There's a gutenberg-mobile PR too: https://github.com/wordpress-mobile/gutenberg-mobile/pull/2744 and there the CI didn't find any issues with package-lock.json changes as well.

etoledom reviewed (approved):
Local E2E tests are ✅ on my machine and CI seems to be happy too 🎉 
Nice job @ceyhun 🙏

guarani reviewed (approved):
It worked after following the tips from https://github.com/WordPress/gutenberg/pull/26283#issuecomment-717172934 (I guess my mistake was only removing the Pods folder instead of `git clean -fdx` to remove _all_ untracked files).
Still on Xcode 12.0 (not 12.1), all tests passed locally! 🎆

ceyhun commented:
That's great, thanks @guarani 🎉 Can you please check the gb-mobile PR too: https://github.com/wordpress-mobile/gutenberg-mobile/pull/2744 and if there aren't any errors, I'll go ahead and merge these two!

ceyhun committed:
Merge branch 'master' into rnmobile/upgrade-ci-to-xcode-12

ceyhun commented:
Merging this despite failing `End-to-End Tests / Admin - 1` as it's failing on current `master` as well.
